File: sdk/js/src/geo.js
/**
 * Calculate the Haversine distance between two points on Earth.
 * @param {number} lat1 - Latitude of point 1
 * @param {number} lon1 - Longitude of point 1
 * @param {number} lat2 - Latitude of point 2
 * @param {number} lon2 - Longitude of point 2
 * @returns {number} Distance in kilometers
 */
export function haversine(lat1, lon1, lat2, lon2) {
  const R = 6371;
  const toRad = (d) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/**
 * Check if an RFQ geographic constraint overlaps a business agent's service area.
 * @param {{ lat: number, lon: number, radius_km: number }} rfqWhere
 * @param {{ lat: number, lon: number, radius_km?: number }} baGeo
 * @returns {boolean}
 */
export function geoMatch(rfqWhere, baGeo) {
  const dist = haversine(rfqWhere.lat, rfqWhere.lon, baGeo.lat, baGeo.lon);
  return dist <= rfqWhere.radius_km + (baGeo.radius_km || 0);
}

File: sdk/python/intentprotocol/geo.py
"""Geographic utilities for Intent Protocol."""

from __future__ import annotations

import math


def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate the great-circle distance between two points on Earth.

    Args:
        lat1, lon1: Coordinates of point 1 (degrees)
        lat2, lon2: Coordinates of point 2 (degrees)

    Returns:
        Distance in kilometers
    """
    R = 6371.0
    to_rad = math.radians

    d_lat = to_rad(lat2 - lat1)
    d_lon = to_rad(lon2 - lon1)

    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(to_rad(lat1)) * math.cos(to_rad(lat2)) * math.sin(d_lon / 2) ** 2
    )
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def geo_match(
    rfq_where: dict,
    ba_geo: dict,
) -> bool:
    """Check if an RFQ location overlaps a business agent's service area.

    Args:
        rfq_where: RFQ geographic constraint {lat, lon, radius_km}
        ba_geo: Business agent geo {lat, lon, radius_km}

    Returns:
        True if areas overlap
    """
    dist = haversine(rfq_where["lat"], rfq_where["lon"], ba_geo["lat"], ba_geo["lon"])
    return dist <= rfq_where.get("radius_km", 5) + ba_geo.get("radius_km", 0)
